homework_02_coroutines

grishan0v opened the pull request:
# рефактор с коллбеков на саспенд функции в презенторе
# добавлен запрос картинок
# реализован viewModel c  CoroutineExceptionHandler

grishan0v committed:
# рефактор с коллбеков на саспенд функции в презенторе
# добавлен запрос картинок
# реализован viewModel c  CoroutineExceptionHandler

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,52 @@
 ) {
 
     private var _catsView: ICatsView? = null
+    private val job = Job()
```
Я думаю лучше здесь использовать [SupervisorJob](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-supervisor-job.html) для еррор процессинга

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,52 @@
-                }
+    private suspend fun getData() {
+        try {
+            val fact = withContext(customPresenterScope.coroutineContext) {
```
Не понял зачем ты берешь контекст родительской корутины и явно указываешь. У тебя же итак унаследуется этот контекст.

antonkazakov replied:
Здесь лучше взять `async` и `Dispatchers.IO`

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,52 @@
+                    checkNotNull(image.body()).file
+                )
+                _catsView?.populate(data)
+            } else CrashMonitor.trackWarning(fact.message())
```
Думаю тут и в тоже нужно какой-то тост показать, иначе пользователю совсем непонятно что произошло

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,52 @@
+
+        } catch (exception: SocketTimeoutException) {
+            _catsView?.showErrorToast("timeout exception")
+        } catch (exception: Exception) {
```
По условию тост тоже нужно показать

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -32,4 +56,10 @@ class CatsPresenter(
     }
+
+    fun stopJob(){
+        job.complete()
```
Здесь лучше вызвать `customPresenterScope.cancel`

antonkazakov commented on app/src/main/java/otus/homework/coroutines/DiContainer.kt:
```diff
@@ -7,10 +7,12 @@ class DiContainer {
 
     val service by lazy { retrofit.create(CatsService::class.java) }
+
+    val viewModel by lazy { MainActivityViewModel(service) }
```
У тебя при использовании `MainActivityViewModel` не падает исключение? Если у `ViewModel` используется non-default конструктор то нужно использовать фабрику https://stackoverflow.com/a/44194893/3416163

grishan0v replied:
Странно, у меня всё работает как часы, а могли это момент поправить? просто я смотрю, что вопросу уже 4 года

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private suspend fun getData() {
+        val fact = withContext(coroutineContext + exceptionHandler) {
```
`coroutineContext` можно не писать, он итак унаследуется. Здесь тебе нужно переключить диспатчер

antonkazakov replied:
`exceptionHandler` можно указать только у родительской корутины

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        val data = CustomCatPresentationModel(
+            checkNotNull(fact.body())[0].fact,
```
Думаю еще нужна обработка кейса что список пустой

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+
+    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        if (e is SocketTimeoutException) _catsData.postValue(Result.Error(e))
+        else CrashMonitor.trackWarning(e.message)
```
_Добавить обработку исключений через try-catch. В случае java.net.SocketTimeoutException показываем Toast с текстом "Не удалось получить ответ от сервером". В остальных случаях логируем исключение в otus.homework.coroutines.CrashMonitor и показываем Toast с exception.message_

В этом кейсе тоже нужно отобразить тост. Иначе пользователю будет непонятно почему ничего не происходит.

antonkazakov reviewed (changes requested):
Привет. Нужно поправить замечания.

grishan0v committed:
# исправление замечаний

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+            _catsData.postValue(Result.Success(data))
+        } else {
+            _catsData.postValue(Result.Error(
+                NullPointerException("data from server was null or empty")
```
Имхо NullPointerException не совсем правильный эксепшен здесь. Я бы просто рантайм выбросил с месседжем, но это мелочи

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+            catsService.getCatImage()
+        }
+
+        if (!fact.body()?.get(0)?.fact.isNullOrEmpty()
```
Очень тяжело читать такую цепочку safe коллов и тело if блока. Это не относится к теме ДЗ, но я бы советовал это переписать)

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,59 @@
+    private suspend fun getData() {
+        try {
+            val fact =
+                withContext(CoroutineScope(Dispatchers.IO).coroutineContext) {
```
Просто `withContext(Dispatchers.IO)`

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,59 @@
-            override fun onResponse(call: Call<Fact>, response: Response<Fact>) {
-                if (response.isSuccessful && response.body() != null) {
-                    _catsView?.populate(response.body()!!)
+    private suspend fun getData() {
```
Имхо можно эту функцию заинлайнить прямо в `onInitComplete`. И еще я бы переименовал `onInitComplete`

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,59 @@
-                if (response.isSuccessful && response.body() != null) {
-                    _catsView?.populate(response.body()!!)
+    private suspend fun getData() {
+        try {
```
Может тогда обработать через эксепшен хендлер?

grishan0v replied:
Вообще можно, просто эксепшен хендлер был в третьем разделе - Реализовать решение ViewModel, а это предыдущая часть задачи связанная с презентором. Ок, добавлю хэндлер

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,59 @@
+                && image.body() != null
+            ) {
+                val data = CustomCatPresentationModel(
+                    checkNotNull(fact.body())[0].fact,
```
А он же по контракту у нас nonnull, разве нет?

grishan0v replied:
да, ты прав

antonkazakov reviewed (changes requested):
Андрей, поправь пожалуйста оставшиеся замечания

grishan0v committed:
# исправление замечаний №2

antonkazakov commented on app/src/main/java/otus/homework/coroutines/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,58 @@
+
+    fun onInitComplete() {
+        viewModelScope.launch(exceptionHandler) {
+            val factResponse = withContext(Dispatchers.IO) {
```
У тебя эти запросы не параллельные. Сначала отработает первый, затем второй. А в данном случае можно распараллелить. Хоть в условиях про параллельность ничего не сказано, я думаю лучше поменять

antonkazakov commented on app/src/main/java/otus/homework/coroutines/DiContainer.kt:
```diff
@@ -7,10 +7,14 @@ class DiContainer {
     val service by lazy { retrofit.create(CatsService::class.java) }
+
+    val viewModel by lazy {
+        MainActivityViewModelFactory(service).create(MainActivityViewModel::class.java)
```
Лучше вьюмодель конечно получать через `ViewModelProviders`

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,28 +1,50 @@
+
+    private suspend fun getData() {
+        customPresenterScope.launch {
+            val fact = withContext(Dispatchers.IO) {
```
Тут тоже лучше распараллелить

antonkazakov reviewed (changes requested):
Андрей привет. Слушай, по всем замечания правки ок, но давай распараллелим запросы?

grishan0v committed:
# исправление замечаний №3
- распараллелил запросы во ViewModel
- распараллелил запросы в Presenter
- сделал инициализацию ViewModel через ViewModelProvider

antonkazakov reviewed (commented):
Супер. Пр закрываю)
